crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -0,0 +1,338 @@
+            native_token,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TradeVerifying for TradeVerifier {
+    async fn verify(
+        &self,
+        query: &PriceQuery,
+        verification: &Verification,
+        trade: Trade,
+    ) -> Result<Estimate> {
+        let solver = dummy_contract!(Solver, trade.solver);
+
+        let settlement = encode_settlement(query, verification, &trade, self.native_token);
+        let settlement =
+            add_balance_queries(settlement, query, verification, self.settlement, &solver);
+
+        let settlement_contract = dummy_contract!(GPv2Settlement, self.settlement);
+        let settlement = settlement_contract
+            .methods()
+            .settle(
+                settlement.tokens,
+                settlement.clearing_prices,
+                settlement.trades,
+                settlement.interactions,
+            )
+            .tx;
+
+        let sell_amount = match query.kind {
+            OrderKind::Sell => query.in_amount.get(),
+            OrderKind::Buy => trade.out_amount,
+        };
+
+        let simulation = solver
+            .methods()
+            .swap(
+                verification.from,
+                query.sell_token,
+                sell_amount,
+                self.native_token,
+                verification.receiver,
+                Bytes(settlement.data.unwrap().0),
+            )
+            .tx;
+
+        let call = CallRequest {
+            // Initiate tx as solver so gas doesn't get deducted from user's ETH.
+            from: Some(solver.address()),
+            to: Some(solver.address()),
+            data: simulation.data,
+            gas: Some(Self::DEFAULT_GAS.into()),
+            ..Default::default()
+        };
+
+        // Set up helper contracts impersonating trader and solver.
+        let mut overrides = hashmap! {
+            verification.from => StateOverride {
+                code: Some(deployed_bytecode!(Trader)),
+                ..Default::default()
+            },
+            solver.address() => StateOverride {
+                code: Some(deployed_bytecode!(Solver)),
+                ..Default::default()
+            },
+        };
+
+        let trader_impl = self.code_fetcher.code(verification.from).await?;
+        if !trader_impl.0.is_empty() {
+            // Store `owner` implementation so `Trader` helper contract can proxy to it.
+            overrides.insert(
+                Self::TRADER_IMPL,
+                StateOverride {
+                    code: Some(trader_impl),
+                    ..Default::default()
+                },
+            );
+        }
+
+        let output = self.simulator.simulate(call, overrides).await?;
+        let summary = SettleOutput::decode(&output)?;
+        let verified = Estimate {
+            out_amount: summary.out_amount(query.kind)?,
+            gas: summary.gas_used.as_u64(),
+            solver: trade.solver,
+            verified: true,
```

Comment:
> Do we even need the Noop one (couldn't we just keep TradeVerifier optional wherever it's used?)

At some point I want to write tests for the logic that prefers verified quotes over unverified ones. Because I think e2e test will be tricky to pull unit test would be the only option.
In order to mock a trade verifier for a unit test I will have to use a trait.
This plus the fact that we eventually want to enforce that we always have a verifier available (and simply don't use it for some quotes) lead to me to model optional trade verification via the NoopVerifier instead of the optional verifier.
When we are ready to always enforce the availability of a trade verifier we can simply drop the NoopVerifier.

> I find it a bit weird that the Noop tradeverifier returns verified: false and the real one returns an error in case it's unable to verify.

Hmm, true. Being able to return an error or `verified: false` is ambiguous. Would be cleaner to return something like `Result<VerifiedQuote, Error>` instead and always return `anyhow!("verification disabled")` from the NoopVerifier.

Will think more about the design.